lib/resources/nginx_conf.rb

```diff
@@ -0,0 +1,64 @@
+# encoding: utf-8
+# author: Dominik Richter
+# author: Christoph Hartmann
+
+require 'utils/nginx_parser'
+
+module Inspec::Resources
+  class NginxConf < Inspec.resource(1)
+    name 'nginx_conf'
+    desc 'Use the nginx_conf InSpec resource to test configuration data '\
+         'for the NginX web server located in /etc/nginx/nginx.conf on '\
+         'Linux and UNIX platforms.'
+    example "
+      describe nginx_conf.params ...
+      describe nginx_conf('/path/to/my/nginx.conf').params ...
+    "
+
+    attr_reader :contents
+
+    def initialize(conf_path = nil)
+      @conf_path = conf_path || '/etc/nginx/nginx.conf'
+      @contents = {}
+    end
+
+    def params
+      @params ||= parse_nginx(@conf_path)
+    end
+
+    def to_s
+      "nginx_conf #{@conf_path}"
+    end
+
+    private
+
+    def read_content(path)
+      return @contents[path] if @contents.key?(path)
+      file = inspec.file(path)
+      if !file.file?
+        return skip_resource "Can't find file \"#{path}\""
+      end
+      @contents[path] = file.content
+    end
+
+    def parse_nginx(path)
+      content = read_content(path)
+      data = NginxConfig.parse(content)
+      resolve_references(data, File.dirname(path))
+    rescue StandardError => e
+      raise "Cannot parse NginX config in #{path}: #{e}"
+    end
+
+    def resolve_references(data, rel_path)
+      return data.map { |x| resolve_references(x, rel_path) } if data.is_a?(Array)
```

Comment:
I've stared at this for a while, but I'm seriously still confused as to what you're trying to accomplish in this method.

 * When will the data passed in be an Array instead of a Hash?
 * In line 54, you return `data` unless it's a Hash, but then the rest of this method does work to make it a hash...

This could really use some comments and examples as to what you're trying to accomplish. I can see this method as a major area of frustration when troubleshooting.